src/main/java/org/opensearch/security/support/Base64Helper.java

```diff
@@ -26,71 +26,131 @@
 
 package org.opensearch.security.support;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InvalidClassException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamClass;
+import java.io.OutputStream;
 import java.io.Serializable;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 
+import com.google.common.base.Preconditions;
+import com.google.common.io.BaseEncoding;
+
+import org.opensearch.OpenSearchException;
+import org.opensearch.SpecialPermission;
+import org.opensearch.core.common.Strings;
+
+import static org.opensearch.security.support.SafeSerializationUtils.isSafeClass;
+
+/**
+ * Provides support for Serialization/Deserialization of objects of supported classes into/from Base64 encoded stream
+ * using JDK's in-built serialization protocol implemented by the ObjectOutputStream and ObjectInputStream classes.
+ */
 public class Base64Helper {
 
-    public static String serializeObject(final Serializable object, final boolean useJDKSerialization) {
-        return useJDKSerialization ? Base64JDKHelper.serializeObject(object) : Base64CustomHelper.serializeObject(object);
+    private final static class SafeObjectOutputStream extends ObjectOutputStream {
+
+        private static final boolean useSafeObjectOutputStream = checkSubstitutionPermission();
+
+        @SuppressWarnings("removal")
+        private static boolean checkSubstitutionPermission() {
+            SecurityManager sm = System.getSecurityManager();
+            if (sm != null) {
+                try {
+                    sm.checkPermission(new SpecialPermission());
+
+                    AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                        AccessController.checkPermission(SUBSTITUTION_PERMISSION);
+                        return null;
+                    });
+                } catch (SecurityException e) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        static ObjectOutputStream create(ByteArrayOutputStream out) throws IOException {
+            try {
+                return useSafeObjectOutputStream ? new Base64Helper.SafeObjectOutputStream(out) : new ObjectOutputStream(out);
+            } catch (SecurityException e) {
+                // As we try to create SafeObjectOutputStream only when necessary permissions are granted, we should
+                // not reach here, but if we do, we can still return ObjectOutputStream after resetting ByteArrayOutputStream
+                out.reset();
+                return new ObjectOutputStream(out);
+            }
+        }
+
+        @SuppressWarnings("removal")
+        private SafeObjectOutputStream(OutputStream out) throws IOException {
+            super(out);
+
+            SecurityManager sm = System.getSecurityManager();
```

Comment:
Let's remove all instances of references to the SM.

Technically we can keep AccessController.doPrivileged but permissions on reflection are not being enforced so it can be removed here safely.